packages/gestalt/src/Masonry/defaultTwoColumnModuleLayout.js

```diff
@@ -117,15 +118,22 @@ function initializeHeightsArray<T>({
     const position = positionCache?.get(item);
     if (position) {
       const col = (position.left - centerOffset) / columnWidthAndGutter;
-      // TODO: Change this to handle multi column
-      const isTwoColumnModule =
-        position.width === calculateMultiColumnModuleWidth(columnWidth, gutter, 2);
-      const heightToAdd = position.height + gutter;
-      heights[col] += heightToAdd;
-      if (isTwoColumnModule) {
-        // if position width is greater than columnWidth
-        // increment height of the neighboring column as well
-        heights[col + 1] += heightToAdd;
+      const columnSpan = getItemColumnSpan(item);
+      const moduleHeight = position.height + gutter;
+      if (position.top < heights[col]) {
+        // handles the case where the column height is already taller because of a previously added multi-column module
+        return;
+      }
+      if (columnSpan > 1) {
+        // find tallest column
+        const colHeights = heights.slice(col, col + columnSpan);
```

Comment:
e.g.
if adding a three column item into col: 3:
- find tallest column between 3, 4, 5
- add module to tallest column
- set heights of other columns to be the same